Discussion: Enable golint?

The current CI indicates that it was desired to run golint, however, in practice it wasn't happening (golint was not present on Travis). It can't be enabled at this moment since it will cause errors (see the list of issues below). If we agree to enable it we will have to clean-up those.

```
cmd/ocagent/config.go:217:5: should replace nonNilExporters += 1 with nonNilExporters++
exporter/trace_exporter.go:28:6: exported type TraceExporter should have comment or be unexported
exporter/trace_exporter.go:36:6: exported type TraceExporterSpanReceiver should have comment or be unexported
exporter/trace_exporter.go:41:1: comment on exported function OCExportersToTraceExporter should be of the form "OCExportersToTraceExporter ..."
exporter/trace_exporter.go:84:1: exported function MultiTraceExporters should have comment or be unexported
exporter/exporterparser/datadog.go:49:1: exported function DatadogTraceExportersFromYAML should have comment or be unexported
exporter/exporterparser/exparser.go:15:1: package comment should be of the form "Package exporterparser ..."
exporter/exporterparser/stackdriver.go:39:1: exported function StackdriverTraceExportersFromYAML should have comment or be unexported
exporter/exporterparser/zipkin.go:31:6: exported type ZipkinConfig should have comment or be unexported
exporter/exporterparser/zipkin.go:60:2: exported const DefaultZipkinEndpointHostPort should have comment (or a comment on this block) or be unexported
exporter/exporterparser/zipkin.go:64:1: exported method ZipkinConfig.EndpointURL should have comment or be unexported
exporter/exporterparser/zipkin.go:73:1: exported function ZipkinExportersFromYAML should have comment or be unexported
interceptor/opencensus/doc.go:15:1: package comment should be of the form "Package ocinterceptor ..."
interceptor/opencensus/opencensus.go:33:6: exported type Interceptor should have comment or be unexported
interceptor/opencensus/opencensus.go:39:1: exported function New should have comment or be unexported
interceptor/opencensus/opencensus.go:54:1: exported method Interceptor.Config should have comment or be unexported
interceptor/opencensus/options.go:19:6: exported type Option should have comment or be unexported
interceptor/opencensus/trace_interceptor.go:32:1: comment on exported function NewTraceInterceptor should be of the form "NewTraceInterceptor ..."
interceptor/opencensus/trace_interceptor.go:37:1: exported function NewTraceInterceptorOnDefaultPort should have comment or be unexported
interceptor/opencensus/trace_interceptor.go:60:10: should omit type error from declaration of var err; it will be inferred from the right-hand side
interceptor/opencensus/trace_interceptor.go:105:10: should omit type error from declaration of var err; it will be inferred from the right-hand side
interceptor/zipkin/trace_interceptor.go:43:6: exported type ZipkinInterceptor should have comment or be unexported
interceptor/zipkin/trace_interceptor.go:50:1: exported function New should have comment or be unexported
interceptor/zipkin/trace_interceptor.go:54:1: exported method ZipkinInterceptor.StartTraceInterception should have comment or be unexported
interceptor/zipkin/trace_interceptor.go:100:1: exported method ZipkinInterceptor.StopTraceInterception should have comment or be unexported
interceptor/zipkin/trace_interceptor_test.go:77:4: should replace nonNilSpans += 1 with nonNilSpans++
interceptor/zipkin/trace_interceptor_test.go:398:3: should replace mp[s] += 1 with mp[s]++
internal/internal.go:88:1: exported function TimeToTimestamp should have comment or be unexported
internal/observability.go:48:5: exported var ViewReceivedSpansInterceptor should have comment or be unexported
internal/observability.go:57:5: exported var ViewExportedSpans should have comment or be unexported
internal/observability.go:65:5: exported var AllViews should have comment or be unexported
internal/version/version.go:26:1: comment on exported var GitHash should be of the form "GitHash ..."
internal/version/version.go:27:13: should omit type string from declaration of var GitHash; it will be inferred from the right-hand side
internal/version/version.go:29:1: exported function Info should have comment or be unexported
metricsreceiver/metricsreceiver.go:29:6: exported type Acknowledgement should have comment or be unexported
spanreceiver/spanreceiver.go:29:6: exported type Acknowledgement should have comment or be unexported
translator/trace/protospan_to_spandata.go:15:1: package comment should be of the form "Package tracetranslator ..."
translator/trace/protospan_to_spandata.go:34:1: exported function ProtoSpanToOCSpanData should have comment or be unexported
translator/trace/protospan_to_spandata.go:247:27: should drop = "" from declaration of var value; it is the zero value
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,4 +6,5 @@ go:
   - 1.10.x
  
 install:
+  - make install-tools
   - go get -t -v ./...

--- a/Makefile
+++ b/Makefile
@@ -3,6 +3,7 @@ ALL_SRC := $(shell find . -type f -name '*.go' -not -path "./vendor/*")
 GOTEST_OPT?=-v -race
 GOTEST=go test
 GOFMT=gofmt
+GOLINT=golint
 GOOS=$(shell go env GOOS)
 
 GIT_SHA=$(shell git rev-parse --short HEAD)
@@ -12,7 +13,7 @@ BUILD_INFO=-ldflags "-X $(BUILD_INFO_IMPORT_PATH).GitHash=$(GIT_SHA)"
 .DEFAULT_GOAL := default_goal
 
 .PHONY: default_goal
-default_goal: fmt test
+default_goal: fmt lint test
 
 .PHONY: test
 test:
@@ -26,6 +27,19 @@ fmt:
 		echo "$$FMTOUT\n"; \
 		exit 1; \
 	fi
+
+.PHONY: lint
+lint:
+	@LINTOUT=`$(GOLINT) ./... 2>&1`; \
+	if [ "$$LINTOUT" ]; then \
+		echo "$(GOLINT) FAILED => clean the following lint errors:\n"; \
+		echo "$$LINTOUT\n"; \
+		exit 1; \
+	fi
+
+.PHONY: install-tools
+install-tools:
+	go get -u golang.org/x/lint/golint
 
 .PHONY: agent
 agent:

--- a/cmd/ocagent/config.go
+++ b/cmd/ocagent/config.go
@@ -214,7 +214,7 @@ func exportersFromYAMLConfig(config []byte) (traceExporters []exporter.TraceExpo
 		for _, te := range tes {
 			if te != nil {
 				traceExporters = append(traceExporters, te)
-				nonNilExporters += 1
+				nonNilExporters++
 			}
 		}
 

--- a/exporter/exporterparser/datadog.go
+++ b/exporter/exporterparser/datadog.go
@@ -46,6 +46,8 @@ type datadogExporter struct {
 	exporter *datadog.Exporter
 }
 
+// DatadogTraceExportersFromYAML parses the yaml bytes and returns an exporter.TraceExporter targeting
+// Datadog according to the configuration settings.
 func DatadogTraceExportersFromYAML(config []byte) (tes []exporter.TraceExporter, doneFns []func() error, err error) {
 	var cfg struct {
 		Exporters *struct {

--- a/exporter/exporterparser/exparser.go
+++ b/exporter/exporterparser/exparser.go
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// This package provides support for parsing and creating the
+// Package exporterparser provides support for parsing and creating the
 // respective exporters given a YAML configuration payload.
 // For now it currently only provides statically imported OpenCensus
 // exporters like:

--- a/exporter/exporterparser/jaeger.go
+++ b/exporter/exporterparser/jaeger.go
@@ -35,7 +35,8 @@ type jaegerExporter struct {
 	exporter *jaeger.Exporter
 }
 
-// JaegerExportersFromYAML reads any configured jaeger exporters from a yaml config
+// JaegerExportersFromYAML parses the yaml bytes and returns an exporter.TraceExporter targeting
+// Jaeger according to the configuration settings.
 func JaegerExportersFromYAML(config []byte) (tes []exporter.TraceExporter, doneFns []func() error, err error) {
 	var cfg struct {
 		Exporters *struct {

--- a/exporter/exporterparser/stackdriver.go
+++ b/exporter/exporterparser/stackdriver.go
@@ -36,6 +36,8 @@ type stackdriverExporter struct {
 
 var _ exporter.TraceExporter = (*stackdriverExporter)(nil)
 
+// StackdriverTraceExportersFromYAML parses the yaml bytes and returns an exporter.TraceExporter targeting
+// Stackdriver according to the configuration settings.
 func StackdriverTraceExportersFromYAML(config []byte) (tes []exporter.TraceExporter, doneFns []func() error, err error) {
 	var cfg struct {
 		Exporters *struct {

--- a/exporter/exporterparser/zipkin.go
+++ b/exporter/exporterparser/zipkin.go
@@ -33,6 +33,7 @@ import (
 	"github.com/census-instrumentation/opencensus-service/exporter"
 )
 
+// ZipkinConfig holds the configuration of a Zipkin exporter.
 type ZipkinConfig struct {
 	ServiceName      string         `yaml:"service_name,omitempty"`
 	Endpoint         string         `yaml:"endpoint,omitempty"`
@@ -56,11 +57,13 @@ type zipkinExporter struct {
 	reporter    zipkinreporter.Reporter
 }
 
+// Default values for Zipkin endpoint.
 const (
 	DefaultZipkinEndpointHostPort = "localhost:9411"
 	DefaultZipkinEndpointURL      = "http://" + DefaultZipkinEndpointHostPort + "/api/v2/spans"
 )
 
+// EndpointURL returns the endpoint URL of the Zipkin configuration.
 func (zc *ZipkinConfig) EndpointURL() string {
 	// If no endpoint was set, use the default Zipkin reporter URI.
 	endpoint := DefaultZipkinEndpointURL
@@ -70,6 +73,8 @@ func (zc *ZipkinConfig) EndpointURL() string {
 	return endpoint
 }
 
+// ZipkinExportersFromYAML parses the yaml bytes and returns an exporter.TraceExporter targeting
+// Zipkin according to the configuration settings.
 func ZipkinExportersFromYAML(config []byte) (tes []exporter.TraceExporter, doneFns []func() error, err error) {
 	var cfg struct {
 		Exporters *struct {

--- a/exporter/trace_exporter.go
+++ b/exporter/trace_exporter.go
@@ -25,6 +25,10 @@ import (
 	tracetranslator "github.com/census-instrumentation/opencensus-service/translator/trace"
 )
 
+// TraceExporter is a interface that receives OpenCensus data, converts it as needed, and
+// sends it to different destinations.
+//
+// ExportSpanData receives OpenCensus data for processing by the exporter.
 type TraceExporter interface {
 	ExportSpanData(ctx context.Context, node *commonpb.Node, spandata ...*trace.SpanData) error
 }
@@ -33,12 +37,15 @@ type toOCExportersTransformer struct {
 	ocTraceExporters []trace.Exporter
 }
 
+// TraceExporterSpanReceiver is a interface connecting a spanreceiver.SpanReceiver and
+// a exporter.TraceExporter. The receiver gets data in different serialization formats,
+// transforms it to OpenCensus in memory data and sends it to the exporter.
 type TraceExporterSpanReceiver interface {
 	TraceExporter
 	spanreceiver.SpanReceiver
 }
 
-// OCExportersToTraceExporterSpanReceiver is a convenience function that transforms
+// OCExportersToTraceExporter is a convenience function that transforms
 // traditional OpenCensus trace.Exporter-s into an opencensus-service TraceExporter.
 // The resulting TraceExporter ignores the passed in node. To make use of the node,
 // please create a custom TraceExporter.
@@ -81,12 +88,14 @@ func (tse *toOCExportersTransformer) ReceiveSpans(ctx context.Context, node *com
 	return ack, err
 }
 
+// MultiTraceExporters wraps multiple trace exporters in a single one.
 func MultiTraceExporters(tes ...TraceExporter) TraceExporterSpanReceiver {
 	return traceExporters(tes)
 }
 
 type traceExporters []TraceExporter
 
+// ExportSpanData exports the span data to all trace exporters wrapped by the current one.
 func (tes traceExporters) ExportSpanData(ctx context.Context, node *commonpb.Node, spandata ...*trace.SpanData) error {
 	for _, te := range tes {
 		_ = te.ExportSpanData(ctx, node, spandata...)
@@ -94,6 +103,8 @@ func (tes traceExporters) ExportSpanData(ctx context.Context, node *commonpb.Nod
 	return nil
 }
 
+// ReceiveSpans receives the span data in the protobuf format, translates it, and forwards the transformed
+// span data to all trace exporters wrapped by the current one.
 func (tes traceExporters) ReceiveSpans(ctx context.Context, node *commonpb.Node, spans ...*tracepb.Span) (*spanreceiver.Acknowledgement, error) {
 	spanDataList := make([]*trace.SpanData, 0, len(spans))
 	for _, span := range spans {

--- a/interceptor/interceptor.go
+++ b/interceptor/interceptor.go
@@ -28,6 +28,11 @@ import (
 //
 // For example it could be Zipkin data source which translates
 // Zipkin spans into *tracepb.Span-s.
+//
+// StartTraceInterception tells the interceptor to start its processing.
+//
+// StopTraceInterception tells the interceptor that should stop interception,
+// giving it a chance to perform any necessary clean-up.
 type TraceInterceptor interface {
 	StartTraceInterception(ctx context.Context, destination spanreceiver.SpanReceiver) error
 	StopTraceInterception(ctx context.Context) error

--- a/interceptor/opencensus/doc.go
+++ b/interceptor/opencensus/doc.go
@@ -12,8 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-/*
-ocinterceptor is the logic for receiving OpenCensus trace protobuf defined spans from
-already instrumented applications and then passing them onto a spanreceiver instance.
-*/
+// Package ocinterceptor is the logic for receiving OpenCensus trace protobuf defined spans from
+// already instrumented applications and then passing them onto a spanreceiver instance.
 package ocinterceptor

--- a/interceptor/opencensus/opencensus.go
+++ b/interceptor/opencensus/opencensus.go
@@ -30,12 +30,14 @@ import (
 	"github.com/census-instrumentation/opencensus-service/spanreceiver"
 )
 
+// Interceptor is the type used to handle spans from OpenCensus exporters.
 type Interceptor struct {
 	spanSink         spanreceiver.SpanReceiver
 	spanBufferPeriod time.Duration
 	spanBufferCount  int
 }
 
+// New creates a new opencensus.Interceptor reference.
 func New(sr spanreceiver.SpanReceiver, opts ...Option) (*Interceptor, error) {
 	if sr == nil {
 		return nil, errors.New("needs a non-nil spanReceiver")
@@ -51,6 +53,7 @@ var _ agenttracepb.TraceServiceServer = (*Interceptor)(nil)
 
 var errUnimplemented = errors.New("unimplemented")
 
+// Config handles configuration messages.
 func (oci *Interceptor) Config(tcs agenttracepb.TraceService_ConfigServer) error {
 	// TODO: Implement when we define the config receiver/sender.
 	return errUnimplemented

--- a/interceptor/opencensus/options.go
+++ b/interceptor/opencensus/options.go
@@ -16,6 +16,9 @@ package ocinterceptor
 
 import "time"
 
+// Option interface defines for configuration settings to be applied to interceptors.
+//
+// WithInterceptor applies the configuration to the given interceptor.
 type Option interface {
 	WithInterceptor(*Interceptor)
 }

--- a/interceptor/opencensus/trace_interceptor.go
+++ b/interceptor/opencensus/trace_interceptor.go
@@ -29,11 +29,12 @@ import (
 	"github.com/census-instrumentation/opencensus-service/spanreceiver"
 )
 
-// New will create a handle that runs an OCInterceptor at the provided address
+// NewTraceInterceptor will create a handle that runs an OCInterceptor at the provided port.
 func NewTraceInterceptor(port int, opts ...Option) (interceptor.TraceInterceptor, error) {
 	return &ocInterceptorHandler{srvPort: port, interceptorOptions: opts}, nil
 }
 
+// NewTraceInterceptorOnDefaultPort will create a handle that runs an OCInterceptor at the default port.
 func NewTraceInterceptorOnDefaultPort(opts ...Option) (interceptor.TraceInterceptor, error) {
 	return &ocInterceptorHandler{srvPort: defaultOCInterceptorPort, interceptorOptions: opts}, nil
 }
@@ -57,7 +58,7 @@ var errAlreadyStarted = errors.New("already started")
 
 // StartTraceInterception starts a gRPC server with an OpenCensus interceptor running
 func (ocih *ocInterceptorHandler) StartTraceInterception(ctx context.Context, sr spanreceiver.SpanReceiver) error {
-	var err error = errAlreadyStarted
+	var err = errAlreadyStarted
 	ocih.startOnce.Do(func() {
 		err = ocih.startInternal(ctx, sr)
 	})
@@ -102,7 +103,7 @@ func (ocih *ocInterceptorHandler) startInternal(ctx context.Context, sr spanrece
 var errAlreadyStopped = errors.New("already stopped")
 
 func (ocih *ocInterceptorHandler) StopTraceInterception(ctx context.Context) error {
-	var err error = errAlreadyStopped
+	var err = errAlreadyStopped
 	ocih.stopOnce.Do(func() {
 		ocih.mu.Lock()
 		defer ocih.mu.Unlock()

--- a/interceptor/zipkin/trace_interceptor.go
+++ b/interceptor/zipkin/trace_interceptor.go
@@ -40,6 +40,7 @@ import (
 	"github.com/census-instrumentation/opencensus-service/spanreceiver"
 )
 
+// ZipkinInterceptor type is used to handle spans received in the Zipkin format.
 type ZipkinInterceptor struct {
 	spanSink spanreceiver.SpanReceiver
 }
@@ -47,10 +48,12 @@ type ZipkinInterceptor struct {
 var _ interceptor.TraceInterceptor = (*ZipkinInterceptor)(nil)
 var _ http.Handler = (*ZipkinInterceptor)(nil)
 
+// New creates a new zipkinginterceptor.ZipkinInterceptor reference.
 func New(sr spanreceiver.SpanReceiver) (*ZipkinInterceptor, error) {
 	return &ZipkinInterceptor{spanSink: sr}, nil
 }
 
+// StartTraceInterception tells the interceptor to start its processing.
 func (zi *ZipkinInterceptor) StartTraceInterception(ctx context.Context, spanSink spanreceiver.SpanReceiver) error {
 	zi.spanSink = spanSink
 	return nil
@@ -97,6 +100,8 @@ func (zi *ZipkinInterceptor) parseAndConvertToTraceSpans(jsonBlob []byte) (reqs 
 	return reqs, nil
 }
 
+// StopTraceInterception tells the interceptor that should stop interception,
+// giving it a chance to perform any necessary clean-up.
 func (zi *ZipkinInterceptor) StopTraceInterception(ctx context.Context) error {
 	return nil
 }

--- a/interceptor/zipkin/trace_interceptor_test.go
+++ b/interceptor/zipkin/trace_interceptor_test.go
@@ -75,7 +75,7 @@ func TestConvertSpansToTraceSpans(t *testing.T) {
 	nonNilSpans := 0
 	for _, span := range req.Spans {
 		if span != nil {
-			nonNilSpans += 1
+			nonNilSpans++
 		}
 	}
 	// Expecting 9 non-nil spans

--- a/internal/internal.go
+++ b/internal/internal.go
@@ -85,6 +85,7 @@ func guessUnixHomeDir() string {
 	return ""
 }
 
+// TimeToTimestamp converts a time.Time to a timestamp.Timestamp pointer.
 func TimeToTimestamp(t time.Time) *timestamp.Timestamp {
 	nanoTime := t.UnixNano()
 	return &timestamp.Timestamp{

--- a/internal/observability.go
+++ b/internal/observability.go
@@ -45,6 +45,7 @@ var itemsDistribution = view.Distribution(
 	100, 150, 200, 250, 300, 450, 500, 600, 700, 800, 900, 1000, 1200, 1400, 1600, 1800, 2000,
 )
 
+// ViewReceivedSpansInterceptor defines the view for the received spans metric.
 var ViewReceivedSpansInterceptor = &view.View{
 	Name:        "oc.io/interceptor/received_spans",
 	Description: "The number of spans received by the interceptor",
@@ -54,6 +55,8 @@ var ViewReceivedSpansInterceptor = &view.View{
 }
 
 var mExportedSpans = stats.Int64("oc.io/interceptor/exported_spans", "Counts the number of exported spans", "1")
+
+// ViewExportedSpans defines the view for exported spans metric.
 var ViewExportedSpans = &view.View{
 	Name:        "oc.io/interceptor/exported_spans",
 	Description: "Tracks the number of exported spans",
@@ -62,6 +65,7 @@ var ViewExportedSpans = &view.View{
 	TagKeys:     []tag.Key{tagKeyExporterName},
 }
 
+// AllViews has the views for the metrics provided by the agent.
 var AllViews = []*view.View{
 	ViewReceivedSpansInterceptor,
 	ViewExportedSpans,

--- a/internal/version/version.go
+++ b/internal/version/version.go
@@ -23,9 +23,11 @@ import (
 // Please increment this value per release.
 const version = "0.0.1"
 
-// This variable will be replaced at link time after `make` has been run.
-var GitHash string = "<NOT PROPERLY GENERATED>"
+// GitHash variable will be replaced at link time after `make` has been run.
+var GitHash = "<NOT PROPERLY GENERATED>"
 
+// Info returns a formatted string, with linebreaks, intended to be displayed
+// on stdout.
 func Info() string {
 	buf := new(bytes.Buffer)
 	rows := [][2]string{

--- a/metricsreceiver/metricsreceiver.go
+++ b/metricsreceiver/metricsreceiver.go
@@ -26,6 +26,8 @@ type MetricsReceiver interface {
 	ReceiveMetrics(ctx context.Context, node *commonpb.Node, metrics ...*metricpb.Metric) (*Acknowledgement, error)
 }
 
+// Acknowledgement struct reports the number of saved and dropped spans in a
+// ReceiveSpans call.
 type Acknowledgement struct {
 	SavedMetrics   uint64
 	DroppedMetrics uint64

--- a/spanreceiver/spanreceiver.go
+++ b/spanreceiver/spanreceiver.go
@@ -26,6 +26,8 @@ type SpanReceiver interface {
 	ReceiveSpans(ctx context.Context, node *commonpb.Node, spans ...*tracepb.Span) (*Acknowledgement, error)
 }
 
+// Acknowledgement struct reports the number of saved and dropped spans in a
+// ReceiveSpans call.
 type Acknowledgement struct {
 	SavedSpans   uint64
 	DroppedSpans uint64

--- a/translator/trace/protospan_to_spandata.go
+++ b/translator/trace/protospan_to_spandata.go
@@ -12,9 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-/*
-Package trace defines translators from Trace proto spans to OpenCensus Go spanData
-*/
+// Package tracetranslator defines translators from Trace proto spans to OpenCensus Go spanData.
 package tracetranslator
 
 import (
@@ -31,6 +29,7 @@ import (
 
 var errNilSpan = errors.New("expected a non-nil span")
 
+// ProtoSpanToOCSpanData transforms a protobuf span into the equivalent trace.SpanData one.
 func ProtoSpanToOCSpanData(span *tracepb.Span) (*trace.SpanData, error) {
 	if span == nil {
 		return nil, errNilSpan
@@ -244,7 +243,7 @@ func protoSpanAttributesToOCAttributes(sa *tracepb.Span_Attributes) map[string]i
 
 	amap := make(map[string]interface{})
 	for pkey, pattr := range sa.AttributeMap {
-		var value interface{} = ""
+		var value interface{}
 
 		if pattr != nil && pattr.Value != nil {
 			switch typ := pattr.Value.(type) {

